xarray/core/dataset.py

```diff
@@ -598,17 +618,34 @@ def __init__(
         if isinstance(coords, Dataset):
             coords = coords.variables
 
-        variables, coord_names, dims, indexes, _ = merge_data_and_coords(
-            data_vars, coords, compat="broadcast_equals"
+        variables, coord_names, dims, ds_indexes, _ = merge_data_and_coords(
+            data_vars,
+            coords,
+            compat="broadcast_equals",
+            create_default_indexes=create_default_indexes,
         )
 
+        both_indexes_and_coords = set(indexes) & coord_names
+        if both_indexes_and_coords:
+            raise ValueError(
+                f"{both_indexes_and_coords} are found in both indexes and coords"
+            )
```

Comment:
This check surprises me. As a user, I would expect that I can write something like `xarray.Dataset(data_vars=ds.data_vars, coords=ds.coords, attrs=ds.attrs, indexes=ds.xindexes)` to make a new object.

Reply:
Good point. Would it be reasonable that coordinates provided via `indexes=...` (if we use it) override coordinates provided via `coords=...`?

Reply:
> Would it be reasonable that the coordinates provided via `indexes=...` (if we use it) override the coordinates provided via `coords=...`?

Can you give an example of how this would happen? Do indexes already provide enough information to create a coordinate (this would slightly surprise me)?

I think an error in these cases or creating a Dataset with inconsistent coordinates/indexes (if it is hard to check) would be fine. I would not silently override coordinates.